spec-insert/lib/doc_processor.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+require 'pathname'
+require_relative 'renderers/spec_insert'
+require_relative 'spec_insert_error'
+
+# Processes a file, replacing spec-insert blocks with rendered content
+class DocProcessor
+  START_MARKER = /<!-- spec_insert_start/
+  END_MARKER = /<!-- spec_insert_end -->/
+
+  def initialize(file_path, logger:)
+    @file_path = Pathname(file_path)
+    @logger = logger
+  end
+
+  # Processes the file, replacing spec-insert blocks with rendered content
+  # @param [Boolean] write_to_file Whether to write the changes back to the file
+  def process(write_to_file: true)
+    relative_path = @file_path.relative_path_from(Pathname.new(Dir.pwd))
+    lines = File.readlines(@file_path)
+    original_content = lines.join
+    insertions = find_insertions(lines)
+    return if insertions.empty?
+
+    insertions.reverse_each { |start, finish, insert| lines[start..finish] = insert.render }
+    rendered_content = lines.join
+    if write_to_file && rendered_content != original_content
+      File.write(@file_path, rendered_content)
+      @logger.info "Spec components inserted into #{relative_path} successfully."
+    end
+    rendered_content
+  rescue SpecInsertError => e
+    @logger.error "Error processing #{relative_path}. #{e.message}"
+  end
+
+  private
+
+  # @return Array<[Integer, Integer, SpecInsert]>
+  def find_insertions(lines)
+    start_indices = lines.each_with_index
+                         .filter { |line, _index| line.match?(START_MARKER) }
+                         .map { |_line, index| index }
+    end_indices = start_indices.map do |index|
+      (index..lines.length - 1).find { |i| lines[i].match?(END_MARKER) }
+    end.compact
+
+    validate_markers(start_indices, end_indices)
+
+    start_indices.zip(end_indices).map do |start, finish|
+      [start, finish, SpecInsert.new(lines[start..finish])]
+    end
+  end
+
+  # @param [Array<Integer>] start_indices
+  # @param [Array<Integer>] end_indices
+  def validate_markers(start_indices, end_indices)
```

Comment:
```suggestion
  def validate_markers!(start_indices, end_indices)
```

Suggest using the Ruby convention of appending a `!` when a method raises.